### PR TITLE
detect: add requires keyword - v2

### PR DESCRIPTION
**Pull request by jasonish**
Add a new rule keyword "requires" that allows a rule to require specific
Suricata versions and/or Suricata features to be enabled.

Example:

  requires: feature geoip, version >= 7.0.0, version < 8;

Feature: https://redmine.openinfosecfoundation.org/issues/5972

Previous PR: https://github.com/OISF/suricata/pull/9915

Changes from previous PR:
- Actually check features
- Pre-scanning

A pre-scan/requires checking mode was adding to the parser. When the requires
flag is set, all the *parsing* for syntax is done, but no validation/setup is
done except for the `requires` and `sid` keywords. This allows `requires` to be
specified after keywords that the engine may not now about yet. However, the
rule still needs to be properly formatted when comes to syntax, quoting,
escaping, etc.

Parsing the SID was also added to the pre-scan, so now any keyword parser
(except requires and sid) can have access to the sid for logging purposes.

The the following rules will now be skipped as they have requirements that are
not satisfied by Suricata:
<pre>
alert vxlan any any -> any any (requires: version >= 10; sid:1;)
alert udp any any -> any any (vxlan_vni:10; requires: version >= 10; sid:2;)
alert http any any => any any (requires: version >= 10; sid:3;)
alert tcp any any -> any any (frame:smtp.not_supported; requires: version >= 10; sid:4;)
</pre>

### CODE REVIEW

**Comment by jasonish** [comment]
Needs S-V tests.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 181 | 193 | 106.63% | 

Pipeline 16795

File: src/detect-parse.c [hunk 1]
```diff
@@ -2144,17 +2160,33 @@ static Signature *SigInitHelper(DetectEngineCtx *de_ctx, const char *sigstr,
...
+    sig->sig_str = SCStrdup(sigstr);
+    if (unlikely(sig->sig_str == NULL)) {
+        goto error;
+    }
+
     /* default gid to 1 */
     sig->gid = 1;
 
-    int ret = SigParse(de_ctx, sig, sigstr, dir, &parser);
+    /* We do a first parse of the rule in a requires, or scan-only
```
Comment (victorjulien) [inline src/detect-parse.c:2168]: how expensive is it to this double parsing? Would it make sense to first do something like a `strstr(sigstr, "requires") != NULL` to skip most of the rules?

File: src/detect-sid.c [hunk 2]
```diff
@@ -67,7 +67,7 @@ static int DetectSidSetup (DetectEngineCtx *de_ctx, Signature *s, const char *si
         SCLogError("sid value 0 is invalid");
         goto error;
     }
-    if (s->id > 0) {
+    if (s->id > 0 && s->id != id) {
```
Comment (victorjulien) [inline src/detect-sid.c:70]: will this also allow `sid:1; sid:1;` in a rule? If so, I'm not really in favor of it.

File: src/detect-parse.c (hunk 1 above)
Comment (victorjulien) [reply src/detect-parse.c:2168]: may be moot if we're also parsing sid in the same pass

**Comment by victorjulien** [comment]
Looking good overall.

**Comment by jasonish** [comment]
Replaced with https://github.com/OISF/suricata/pull/9947.

**Pull request closed without merging**